app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java

```diff
@@ -0,0 +1,544 @@
+                    }
+                    cur.moveToNext();
+                }
+            }
+            return recordObjects;
+        } finally {
+            if (cur != null) {
+                cur.close();
+            }
+        }
+    }
+
+    private T readObjectFromFile(Cursor cursor, int dbEntryId) {
+        return readObjectFromFile(cursor,
+                cursor.getColumnIndexOrThrow(DatabaseHelper.FILE_COL),
+                cursor.getColumnIndexOrThrow(DatabaseHelper.AES_COL),
+                dbEntryId);
+    }
+
+    private T readObjectFromFile(Cursor cursor, int fileColIndex,
+                                 int aesColIndex, int dbEntryId) {
+        String filename = cursor.getString(fileColIndex);
+        byte[] aesKeyBlob = cursor.getBlob(aesColIndex);
+        InputStream inputStream = null;
+        try {
+            inputStream = getInputStreamFromFile(filename, aesKeyBlob);
+            return newObject(inputStream, dbEntryId);
+        } finally {
+            if (inputStream != null) {
+                try {
+                    inputStream.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+    }
+
+    private SQLiteDatabase getDbOrThrow() {
+        try {
+            return helper.getHandle();
+        } catch (SessionUnavailableException e) {
+            throw new UserStorageClosedException(e.getMessage());
+        }
+    }
+
+    protected InputStream getInputStreamFromFile(String filename, byte[] aesKeyBytes) {
+        SecretKeySpec aesKey = new SecretKeySpec(aesKeyBytes, "AES");
+        return EncryptionIO.getFileInputStream(filename, aesKey);
+    }
+
+    @Override
+    public T getRecordForValues(String[] rawFieldNames, Object[] values)
+            throws NoSuchElementException, InvalidIndexException {
+        SQLiteDatabase db = getDbOrThrow();
+
+        Pair<String, String[]> whereArgsAndVals =
+                helper.createWhereAndroid(rawFieldNames, values, em, null);
+        Cursor cur = db.query(table, dataColumns,
+                whereArgsAndVals.first, whereArgsAndVals.second,
+                null, null, null);
+        try {
+            int queryCount = cur.getCount();
+            if (queryCount == 0) {
+                throw new NoSuchElementException("No element in table " + table +
+                        " with names " + Arrays.toString(rawFieldNames) +
+                        " and values " + Arrays.toString(values));
+            } else if (queryCount > 1) {
+                throw new InvalidIndexException("Invalid unique column set" +
+                        Arrays.toString(rawFieldNames) +
+                        ". Multiple records found with value " +
+                        Arrays.toString(values), Arrays.toString(rawFieldNames));
+            }
+            cur.moveToFirst();
+            byte[] serializedObj = cur.getBlob(cur.getColumnIndexOrThrow(DatabaseHelper.DATA_COL));
```

Comment:
The logic is slightly different in each case:
 - in `getRecordsForValues` I would guess we want to compute the column indexes outside of the `while` loop
 - in `getRecordForValues` we avoid loading the entire byte array by using an input stream
 - while in `readBytes` we directly read out of the file into a byte array.

They could all be unified but I suspect it would lead to slight performance losses. Your call since you understand these implications better than myself.